README
The InteractiveTimeline extension is a MediaWiki Tag extension that allows
users to insert timelines into MediaWiki pages. The timelines are shown
using the CHAP Timeline javascript library developed by Almende, so they
may be panned and zoomed by the user, and events in the timeline may
contain images, links to other wiki pages or sites, and other formatting.

Full documentation may be found on the [extension page](https://www.mediawiki.org/wiki/Extension:InteractiveTimeline)


CHANGES
Version 1.9 (Canada Water) 2019-05-11

    [BUG FIXES]

    * Replaced CHAP links library with Vis.js loaded from CDN

Version 1.6 (Moorgate) 2016-07-23

    [ENHANCEMENTS]

    * Updated structure to conform to 1.25+ extension registration system.

Version 1.5 (Baker Street) 2015-03-26

    [BUG FIXES]

    * Moved itimeline configuration data out of mw.config (Issue #4)

Version 1.4 (Temple) 2015-03-25

    [ENHANCEMENTS]

    * Allow groups to contain simple markup (images, formatting, links)

Version 1.3 (Warren Street) 2015-03-25

    [ENHANCEMENTS]

    * Enable support for grouping in timelines. (Issue #3, Luis Miguel Morillas)

Version 1.2 (Lancaster Gate) 2015-01-19

    [BUG FIXES]

    * Fix handling of date options (min, max, start, end) (Issue #2, knyazhe1)

Version 1.1 (Stamford Brook) 2014-07-13

    [BUG FIXES]

    * Disallow 24:00:00 time in events (Date.parse() failure in all the browsers)
    * Stupid bugs in event start/end date parsing

Version 1.0 (North Ealing) 2014-07-12

    Initial release.

InteractiveTimeline.php
<?php
if ( ! defined( 'MEDIAWIKI' ) )
		die();
/**
 * A parser extension to MediaWiki that adds the <itimeline> tag to
 * allow interactive timelines to be displayed via the CHAP Timeline
 * library developed by Almende B.V.
 *
 * @section LICENSE
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @ingroup Extensions
 * @author Chris Page <chris@starforge.co.uk>
 * @copyright Copyright © 2014-2016 Chris Page
 * @license GNU General Public Licence 2.0 or later
 */

if ( function_exists( 'wfLoadExtension' ) ) {
	wfLoadExtension( 'InteractiveTimeline' );
	// Keep i18n globals so mergeMessageFileList.php doesn't break
	$wgMessagesDirs['InteractiveTimeline'] = __DIR__ . '/i18n';
	$wgExtensionMessagesFiles['InteractiveTimelineAlias'] = __DIR__ . '/InteractiveTimeline.alias.php';
	wfWarn(
		'Deprecated PHP entry point used for InteractiveTimeline extension. ' .
		'Please use wfLoadExtension instead, ' .
		'see https://www.mediawiki.org/wiki/Extension_registration for more details.'
	);
	return;
} else {
	die( 'This version of the InteractiveTimeline extension requires MediaWiki 1.25+' );
}


InteractiveTimeline.i18n.php
<?php
/**
 * This is a backwards-compatibility shim, generated by:
 * https://git.wikimedia.org/blob/mediawiki%2Fcore.git/HEAD/maintenance%2FgenerateJsonI18n.php
 *
 * Beginning with MediaWiki 1.23, translation strings are stored in json files,
 * and the EXTENSION.i18n.php file only exists to provide compatibility with
 * older releases of MediaWiki. For more information about this migration, see:
 * https://www.mediawiki.org/wiki/Requests_for_comment/Localisation_format
 *
 * This shim maintains compatibility back to MediaWiki 1.17.
 */
$messages = array();
if ( !function_exists( 'wfJsonI18nShima4150441b06af5f5' ) ) {
	function wfJsonI18nShima4150441b06af5f5( $cache, $code, &$cachedData ) {
		$codeSequence = array_merge( array( $code ), $cachedData['fallbackSequence'] );
		foreach ( $codeSequence as $csCode ) {
			$fileName = dirname( __FILE__ ) . "/i18n//$csCode.json";
			if ( is_readable( $fileName ) ) {
				$data = FormatJson::decode( file_get_contents( $fileName ), true );
				foreach ( array_keys( $data ) as $key ) {
					if ( $key === '' || $key[0] === '@' ) {
						unset( $data[$key] );
					}
				}
				$cachedData['messages'] = array_merge( $data, $cachedData['messages'] );
			}

			$cachedData['deps'][] = new FileDependency( $fileName );
		}
		return true;
	}

	$GLOBALS['wgHooks']['LocalisationCacheRecache'][] = 'wfJsonI18nShima4150441b06af5f5';
}


modules/ext.interactivetimeline.js
/**
 * Main javascript file for the Interactive Timeline extension. This file
 * is needed to instantiate vis.Timeline objects wherever itimeline
 * divs are found on a page, converting the list of events in the div
 * into a format Timeline can understand.
 *
 * @author Chris Page <chris@starforge.co.uk>
 */
// Note the leading ; is deliberate.
;(function( mw, $ ) {

	$.InteractiveTimeline = function ( el ) {
		var base = this;

		// Access to jQuery and DOM versions of element
		base.$el = $( el );
		base.el = el;

		// Add a reverse reference to the DOM object
		base.$el.data( "InteractiveTimeline", base );

		base.init = function() {
			base.config = fetchConfig(base.$el);

			// extract the data from the element
			base.data = buildData( base.$el );

			// Build the new timeline
			base.timeline = new vis.Timeline( base.el );
			base.timeline.setOptions(base.config);
			if(base.data.groups) base.timeline.setGroups(base.data.groups);
			base.timeline.setItems(base.data.items);
		};

		// Run initializer
		base.init();
	};

	/**
	 * Convert the itl-event children of the specified container to an
	 * array of event objects suitable to pass to vis.Timeline.
	 *
	 * @param {Object} container The jQuery element containing the event data.
	 * @return {Array} An array of event objects.
	 */
	function buildData( container ) {
		// Get the list of itl-events defined for this timeline container
		var events = container.children( 'div.itl-event' );
		var data   = {};
		data.items = new vis.DataSet();
		var groups = new vis.DataSet();
		var groupc = 0;
		var grouphash = {};

		// Process each event into an object timeline can understand.
		events.each( function( index, elem ) {
			var jElem = $( elem );
			// Fetch the elements that contain the start, end, and body if possible
			var startdate = jElem.children( 'div.itl-start' ).text();
			var enddate = jElem.children( 'div.itl-end' ).text();
			var body = jElem.children( 'div.itl-body' ).html();
			var group = jElem.children( 'div.itl-group' ).html();

			// If a group has been specified, convert it to an ID.
			if(group !== undefined) {
				if(!grouphash[group]) {
					grouphash[group] = ++groupc;
				}

				group = grouphash[group];
			}

			// Must have a start date and body element.
			if ( startdate && body ) {
				var event = { 'start': new Date( startdate ),
							  'content': body,
							};
				if(group !== undefined) event.group = group

				// If an end date has be set, store that too.
				if ( enddate) {
					event.end = new Date( enddate )
				}

				data.items.add( event );
			}
		});

		// If we have any groups, build a group data set
		if(Object.keys(grouphash).length) {
			$.each(grouphash, function (key, val) {
				groups.add({ "id": val,
							 "content": key
						   });
			});

			data.groups = groups;
		}

		return data;
	};

	/** Read the configuration for the timeline from the itl-config div.
	 * This parses the timeline configuration out of the config div, and
	 * returns an object containing the settings.
	 *
	 * @param {Object} element The jQuery element for the timeline.
	 * @return {Object} An object containing the timeline settings.
	 */
	function fetchConfig( element ) {
		var rawconfig = element.children( 'div.itl-config' ).text();

		if ( rawconfig ) {
			var config = JSON.parse( rawconfig );

			// Fix up options that need to be in Date form
			return convertDateOptions(config);
		}

		return { };
	}

	/** Given an objet containing Timeline options, convert any options that
	 *  should be Date objects from the string representation used in the
	 *  serialised options into Date objects.
	 *
	 * @param {Object} options The object containing the options to process.
	 */
	function convertDateOptions( options ) {
		// The names of the options that should be Date objects
		var dateOptions = ['min', 'max', 'start','end'];

		// Options are optional, so only process the object if any are set.
		if ( options ) {
			// Convert the arguments if needed
			dateOptions.forEach( function( opt ) {
				if ( options[opt] != null && typeof options[opt] !== "Date" ) {
					options[opt] = new Date( options[opt] );
				}
			});
		}

		return options;
	};

	// convert all itimeline div elements to timelines.
	$( document ).ready( function() {
		$( '.itimeline' ).each( function() {
			new $.InteractiveTimeline( this );
		});
	});

})( mediaWiki, jQuery );


InteractiveTimeline.body.php
<?php
/**
 * This file is part of the InteractiveTimeline Extension to MediaWiki
 * https://www.mediawiki.org/wiki/Extension:InteractiveTimeline
 *
 * @section LICENSE
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @ingroup Extensions
 * @author Chris Page <chris@starforge.co.uk>
 * @copyright Copyright © 2014-2016 Chris Page
 * @license GNU General Public Licence 2.0 or later
 */

class InteractiveTimeline {

	/**
	 * Determine whether the specified argument is a valid boolean value.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validBoolean( $arg, &$valid ) {
		$arg = trim( $arg );

		if ( preg_match( "/^(true|false)$/i", $arg, $matches ) ) {
			$valid = true;
			return (strtolower($matches[0]) === 'true');
		}

		$valid = false;
		return null;
	}


	/**
	 * Determine whether the specified argument represents a valid css size.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validCSSSize( $arg, &$valid ) {
		$arg = trim( $arg );

		if ( preg_match( "/^(-?\d+(\.\d+)?(%|cm|mm|in|em|ex|pt|pc|px)?)$/i", $arg, $matches ) ) {
			$valid = true;
			return $matches[0];
		}

		$valid = false;
		return null;
	}


	/**
	 * Determine whether the specified argument is a valid date with optional time.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validDatetime( $arg, &$valid ) {
		$arg = trim( $arg );

		// Note that this is not a full IS8601 checker - it does not support periods,
		// ordinal dates, or decimals in the time section. It also only supports hours
		// in the range 00-23 (ISO8601 allows 24) as *none* of the Date.parse() implementations
		// in any browser I've checked will parse '24:00:00' correctly.
		// $matches[N]       1         2                 3                         4                 5              6           7      8               9
		//                   YYYY  -  MM            -  DD(restricted)          T   HH               :MM            :SS          Z +/-  HH             :MM
		if ( preg_match( "/^(\d{4})-?(0[1-9]|1[0-2])-?(0[1-9]|[12]\d|3[01])(?:[ T]([01]\d|2[0-3])(?::([0-5]\d))?(?::([0-5]\d))?([-+Z])(0\d|1[0-4])?(?::?([0-5]\d))?)?$/", $arg, $matches)){
			$valid = true;

			// Date must be specified in full
			$date = $matches[1] . "-" . $matches[2] . "-" . $matches[3] . "T";

			// Hour, minute, and second are optional and should default to zero
			$date .= ( isset( $matches[4] ) && $matches[4] !== '' ? $matches[4] : "00" ) . ":" .
				( isset( $matches[5] ) && $matches[5] !== '' ? $matches[5] : "00" ) . ":" .
				( isset( $matches[6] ) && $matches[6] !== '' ? $matches[6] : "00" );

			// If a timezone is set, use it
			if ( isset( $matches[7] ) && ( $matches[7] === 'Z' || isset( $matches[8] ) ) ) {
				$date .= $matches[7];

				// If the time is not in UTC, and offset must be specified.
				if ( $matches[7] !== 'Z' ) {
					$date .= $matches[8];

					// Minute part of time offset is optional
					$date .= ":" . ( isset( $matches[9] ) && $matches[9] !== '' ? $matches[9] : "00" );
				}

			// Otherwise default to UTC explicitly (if not included the browser
			// will use local time)
			} else {
				$date .= "Z";
			}

			return $date;
		}

		$valid = false;
		return null;
	}


	/**
	 * Determine whether the specified argument is a valid integer value.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validInteger( $arg, &$valid ) {
		$arg = trim( $arg );

		if ( preg_match( "/^\d+$/", $arg, $matches) ) {
			$valid = true;
			return intval( $matches[0] );
		}

		$valid = false;
		return null;
	}


	/**
	 * Determine whether the specified argument is a locale supported by
	 * the CHAP Timeline code. This expects timeline-locales.js to be loaded.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validLocale( $arg, &$valid ) {
		$arg = trim( $arg );

		if ( preg_match( "/^(ca(?:_ES)?|en(?:_US|_UK)?|nl(?:_NL|_BE)?|fi(?:_FI)?|fr(?:_FR|_BE|_CA)?|de(?:_DE|_CH)?|da(?:_DK)?|ru(?:_RU)?|es(?:_ES)?|tr(?:_TR)?)$/", $arg, $matches ) ) {
			$valid = true;
			return $matches[0];
		}

		$valid = false;
		return null;
	}


	/**
	 * Determine whether the specified argument is a valid CHAP Timeline
	 * style value. This only accepts the built-in styles and it does
	 * not support custom styles.
	 *
	 * @param string $arg    The value to check.
	 * @param boolean $valid A reference to a variable that will be set to true if
	 *                       the argument is valid, false if it is not.
	 * @return string The validated value or null if validation failed
	 */
	public static function validTimestyle( $arg, &$valid ) {
		$arg = trim( $arg );

		if ( preg_match( "/^(box|dot)$/i", $arg, $matches ) ) {
			$valid = true;
			return strtolower( $matches[0] );
		}

		$valid = false;
		return null;
	}


	/**
	 * Validate the argument with the specified name, and store the validated
	 * result in the options array. If no value has been set for the named
	 * argument, or the value specified for the argument is not valid, this
	 * will not update the options array.
	 *
	 * @param array $options A reference to an array to store the validated
	 *                         options in.
	 * @param array $args     A reference to an array containing the arguments
	 *                         supplied by the tag.
	 * @param string $name     The name of the argument to validate. This may
	 *                         contain mixed case: even though MediaWiki will
	 *                         lowercase names in the args array, the options may
	 *                         require mixed case. The name supplied is converted
	 *                         to lowercase to look up the value in args, and
	 *                         used 'as is' when setting the value in options.
	 * @param string $type     The type of validation to apply to the value.
	 *                         Supported values are 'boolean', 'csssize',
	 *                         'datetime', 'integer', 'locale', and 'timestyle'.
	 */
	public static function validateArgument( &$options, &$args, $name, $type) {
		// All mediawiki args are lowercase, but the options may be lowerCamelCase.
		$lcname = strtolower( $name );

		// Only bother trying to do anything if the user has set the option
		if ( isset( $args[$lcname] ) ) {
			$value = null;

			// Use the appropriate validator to check the option
			switch ( $type ) {
				case "boolean": $value = self::validBoolean( $args[$lcname], $valid );
					break;
				case "csssize": $value = self::validCSSSize( $args[$lcname], $valid );
					break;
				case "datetime": $value = self::validDatetime( $args[$lcname], $valid );
					break;
				case "integer": $value = self::validInteger( $args[$lcname], $valid );
					break;
				case "locale": $value = self::validLocale( $args[$lcname], $valid );
					break;
				case "timestyle": $value = self::validTimestyle( $args[$lcname], $valid );
					break;
			}

			// If a valid value was obtained, store it.
			if ( isset( $value ) && $valid ) {
				$options[$name] = $value;
			}
		}
	}


	/**
	 * Construct the options to set for the timeline. This validates the arguments
	 * set in the itimeline tag and stores valid arguments in the options array.
	 *
	 * @param array $options A reference to an array to store the validated options.
	 * @param array $args    A reference to an array containing the arguments
	 *                       supplied by the tag.
	 */
	public static function buildTimelineOptions( &$options, &$args ) {

		// Establish any defaults that differ from timeline's own defaults
		$options['selectable']= false;      // No point in making timeline entries selectable

		// And now check any user-specified arguments
		self::validateArgument( $options, $args, 'end', 'datetime' );
		self::validateArgument( $options, $args, 'height', 'csssize' );
		self::validateArgument( $options, $args, 'locale', 'locale' );
		self::validateArgument( $options, $args, 'max', 'datetime' );
		self::validateArgument( $options, $args, 'min', 'datetime' );
		self::validateArgument( $options, $args, 'minHeight', 'integer' );
		self::validateArgument( $options, $args, 'moveable', 'boolean' );
		self::validateArgument( $options, $args, 'stack', 'boolean' );
		self::validateArgument( $options, $args, 'start', 'datetime' );
		self::validateArgument( $options, $args, 'style', 'timestyle' );
		self::validateArgument( $options, $args, 'showCurrentTime', 'boolean' );
		self::validateArgument( $options, $args, 'showMajorLabels', 'boolean' );
		self::validateArgument( $options, $args, 'showMinorLabels', 'boolean' );
		self::validateArgument( $options, $args, 'width', 'csssize' );
		self::validateArgument( $options, $args, 'zoomable', 'boolean' );
		self::validateArgument( $options, $args, 'zoomMax', 'integer' );
		self::validateArgument( $options, $args, 'zoomMin', 'integer' );
	}


	/**
	 * Given a timeline event definition line, ensure that it contains the required
	 * parts (start date and event description, or start, end, and description) and
	 * that the start and possibly end values are valid dates. This generates the
	 * divs that allow easier extraction of the event information in javascript.
	 *
	 * @param string $line The line containing the timeline event to validate.
	 * @return string The divs describing the event if the line is valid, null otherwise.
	 */
	public static function buildTimelineLine( $line ) {

		// Sections are delimited by |
		$parts = explode( "|", trim( $line ) );

		// Lines *must* consist of two or three parts: a date or interval, an optional group, and the text
		if ( count( $parts ) == 2 || count( $parts ) == 3 ) {

			// The first part might be an interval
			$dates = explode( "/", $parts[0]);

			// The first part must be a datetime string, or the whole line is rejected
			$value = self::validDatetime( $dates[0], $valid );
			if ( $valid ) {
				$output = Html::element( 'div', array( 'class' => 'itl-start' ), $value );

				// If there are two dates, the second is the end date
				if ( count( $dates ) == 2) {
					$value = self::validDatetime( $dates[1], $valid );
					if ( $valid ) {
						$output .= Html::element( 'div', array( 'class' => 'itl-end' ), $value );

					}
				}

				// Two parts implies date and text...
				if ( count( $parts ) == 2 ) {
					$body = $parts[1];

				// While three is date, group, and text
				} else {
					$body = $parts[2];
					$output .= Html::rawelement( 'div', array( 'class' => 'itl-group' ), $parts[1] );
				}
				$output .= Html::rawelement( 'div', array( 'class' => 'itl-body' ), $body );


				return $output;
			}
		}

		return null;
	}


	/**
	 * Validate the timeline events in the itimeline tag body, and produce a series
	 * of divs containing the data, one per event, for easier parsing in javascript.
	 *
	 * @param string $input   The content of the tag.
	 * @param array $args     The attributes of the tag.
	 * @param Parser $parser  Parser instance available to render wikitext into html,
	 *                        or parser methods.
	 * @param PPFrame $frame  Can be used to see what template arguments ({{{1}}})
	 *                        this hook was used with.
	 * @return string The itl-event list to use inside the itimeline tag.
	 */
	public static function buildTimelineEvents( $input, $parser, $frame ) {

		// First expand any templates/transclusions
		$body = $parser->recursiveTagParse( $input, $frame );

		// now split on lines
		$lines = explode( "\n", $body );

		$output = "";
		// Convert the lines to validated output
		foreach ( $lines as $line ) {
			$linedata = self::buildTimelineLine( $line );
			if ( isset( $linedata ) ) {
				$output .= Html::rawelement( 'div', array( 'class' => 'itl-event' ), $linedata ) . "\n";
			}
		}

		return $output;
	}


	/**
	 * Parser hook handler for <itimeline>
	 *
	 * @param string $input   The content of the tag.
	 * @param array $args     The attributes of the tag.
	 * @param Parser $parser  Parser instance available to render wikitext into html,
	 *                        or parser methods.
	 * @param PPFrame $frame  Can be used to see what template arguments ({{{1}}})
	 *                        this hook was used with.
	 * @return string HTML to insert in the page.
	 */
	public static function parserHook( $input, $args = array(), $parser, $frame ) {
		global $wgOut;

		static $tlNumber = 0;
		$elemID = 'itimeline-' . ++$tlNumber;

		$options = array();
		self::buildTimelineOptions( $options, $args );

		$timelinedata = self::buildTimelineEvents( $input, $parser, $frame );

		// Store the timeline setup
		$timelinedata .=  Html::rawelement( 'div', array( 'class' => 'itl-config' ), FormatJson::encode( $options ) );

		return Html::rawelement( 'div', array( 'id' => $elemID, 'class' => 'itimeline' ), $timelinedata );
	}


	/**
	 * Add the Interactive Timeline resource modules to the load queue of all pages.
	 *
	 * @param OutputPage $out Instance of OutputPage.
	 * @param Skin $skin The skin instance.
	 * @return boolean Always returns true.
	 */
	public static function onBeforePageDisplay( &$out, &$skin ) {
		global $wgITvisjsCDNcss, $wgITvisjsCDNjs;

		// Ensure that the required resource modules are loaded
		$out->addModules( 'ext.InteractiveTimeline.loader' );

		// Load vis.js from a CDN.
		$script = '
			<link rel="stylesheet" type="text/css" href="'.$wgITvisjsCDNcss.'">
			<script type="text/javascript" src="'.$wgITvisjsCDNjs.'"></script>
		';

		$out->addHeadItem("InteractiveTimeline CDN", $script);

		return true;
	}


	/**
	 * Register the <itimeline> tag with the Parser.
	 *
	 * @param $parser Parser instance of Parser
	 * @return boolean Always returns true
	 */
	public static function onParserFirstCallInit( &$parser ) {
		// Adds the <itimeline>...</itimeline> tag to the parser.
		$parser -> setHook( 'itimeline', 'InteractiveTimeline::parserHook' );

		return true;
	}
}
